File: backend/api/v1/pipeline.py
"""Signal Pipeline API — scheduler control and pipeline status.

Admin endpoints for managing the signal acquisition pipeline.
"""

import logging

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from backend.auth.dependencies import get_current_user
from backend.auth.guards import require_role
from backend.auth.schemas import AuthContext
from backend.database import get_db
from backend.job_queue import enqueue_job, get_queue_stats
from backend.repositories.signal_contract import SignalContractRepository
from backend.schemas.signals import FetchTierRequest, PipelineStatusResponse

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/pipeline")


@router.get("/status", response_model=PipelineStatusResponse)
async def get_pipeline_status(
    auth: AuthContext = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Get signal acquisition pipeline status."""
    from backend.signals.scheduler import get_scheduler

    scheduler = get_scheduler()
    repo = SignalContractRepository(db)

    active = await repo.get_active_contracts(org_id=auth.org_id)
    degraded = await repo.get_degraded_contracts(org_id=auth.org_id)

    return PipelineStatusResponse(
        scheduler_running=scheduler.is_running,
        active_contracts=len(active),
        degraded_contracts=len(degraded),
        degraded_names=[c.name for c in degraded],
    )


@router.post("/fetch")
async def trigger_tier_fetch(
    body: FetchTierRequest,
    auth: AuthContext = Depends(get_current_user),
):
    """Manually trigger a fetch for all contracts in a tier.

    Enqueues an RQ job — does not block. Requires admin role.
    """
    require_role(auth, "admin")
    from backend.jobs.acquisition_job import fetch_signals_by_tier

    queue_name = "high" if body.tier == "realtime" else "default"
    job = enqueue_job(
        fetch_signals_by_tier,
        body.tier,
        queue_name=queue_name,
        job_timeout="15m",
    )
    return {
        "status": "queued",
        "job_id": job.id,
        "tier": body.tier,
    }


@router.get("/queues")
async def get_pipeline_queues(
    auth: AuthContext = Depends(get_current_user),
):
    """Get RQ queue statistics."""
    return get_queue_stats()


@router.post("/scheduler/start")
async def start_scheduler(
    auth: AuthContext = Depends(get_current_user),
):
    """Start the signal acquisition scheduler. Requires admin role."""
    require_role(auth, "admin")
    from backend.signals.scheduler import get_scheduler

    scheduler = get_scheduler()
    if scheduler.is_running:
        return {"status": "already_running"}

    scheduler.start()
    return {"status": "started"}


@router.post("/scheduler/stop")
async def stop_scheduler(
    auth: AuthContext = Depends(get_current_user),
):
    """Stop the signal acquisition scheduler. Requires admin role."""
    require_role(auth, "admin")
    from backend.signals.scheduler import get_scheduler

    scheduler = get_scheduler()
    if not scheduler.is_running:
        return {"status": "not_running"}

    scheduler.stop()
    return {"status": "stopped"}


@router.get("/source-health")
async def get_source_health(
    auth: AuthContext = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Get health status of all active signal contracts.

    Returns summary with healthy/stale/degraded/critical counts
    and lists of problematic contracts. Requires admin role.
    """
    require_role(auth, "admin")
    from backend.services.source_health import SourceHealthService

    service = SourceHealthService(db)
    return await service.get_health_summary(org_id=auth.org_id)


@router.get("/source-health/{contract_id}")
async def get_contract_health(
    contract_id: str,
    auth: AuthContext = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Get detailed health for a specific signal contract.

    Includes recent signal delivery counts and freshness metrics.
    Requires admin role.
    """
    require_role(auth, "admin")
    from uuid import UUID

    from backend.services.source_health import SourceHealthService

    service = SourceHealthService(db)
    result = await service.get_contract_health(UUID(contract_id), org_id=auth.org_id)
    if not result:
        from fastapi import HTTPException

        raise HTTPException(status_code=404, detail="Contract not found")
    return result

File: frontend/lib/api/pipeline.ts
/**
 * Pipeline API service.
 *
 * Maps to: backend/api/v1/pipeline.py
 */

import { get, post } from './client';
import type {
  FetchTierRequest,
} from './types';

export interface PipelineAdminStatusResponse {
  scheduler_running: boolean;
  active_contracts: number;
  degraded_contracts: number;
  degraded_names: string[];
  queues: Record<string, { name: string; count: number; failed: number; scheduled: number }>;
  workers_online: number;
  workers: Array<{
    name: string;
    state: string;
    queues: string[];
    current_job_id: string | null;
    last_heartbeat: string | null;
  }>;
  provider_readiness: Record<string, boolean>;
}

/** GET /api/v1/pipeline/status */
export function getPipelineStatus() {
  return get<PipelineAdminStatusResponse>('/pipeline/status');
}

/** POST /api/v1/pipeline/fetch */
export function triggerTierFetch(body: FetchTierRequest) {
  return post<{ status: string; job_id: string }>('/pipeline/fetch', body);
}

/** GET /api/v1/pipeline/queues */
export function getPipelineQueues() {
  return get<Record<string, unknown>>('/pipeline/queues');
}

/** POST /api/v1/pipeline/scheduler/start */
export function startScheduler() {
  return post<{ status: string }>('/pipeline/scheduler/start');
}

/** POST /api/v1/pipeline/scheduler/stop */
export function stopScheduler() {
  return post<{ status: string }>('/pipeline/scheduler/stop');
}

// ── Source Health ──────────────────────────────────────────────────────────

export interface SourceHealthContract {
  id: string;
  name: string;
  source_url: string;
  source_type: string;
  schedule_tier: string;
  status: string;
  failure_count: number;
  last_fetched_at: string | null;
  last_error: string | null;
  health: 'healthy' | 'stale' | 'degraded' | 'critical';
  is_auto_discovered: boolean;
}

export interface SourceHealthSummary {
  total_active: number;
  healthy: number;
  stale: number;
  degraded: number;
  critical: number;
  stale_contracts: SourceHealthContract[];
  degraded_contracts: SourceHealthContract[];
  critical_contracts: SourceHealthContract[];
  auto_discovered_active: number;
}

export interface ContractDetailedHealth extends SourceHealthContract {
  max_failures: number;
  signals_24h: number;
  signals_7d: number;
  signals_30d: number;
  last_signal_at: string | null;
  hours_since_signal: number | null;
}

/** GET /api/v1/pipeline/source-health */
export function getSourceHealth() {
  return get<SourceHealthSummary>('/pipeline/source-health');
}

/** GET /api/v1/pipeline/source-health/:contractId */
export function getContractHealth(contractId: string) {
  return get<ContractDetailedHealth>(`/pipeline/source-health/${contractId}`);
}
